l1-contracts/src/core/Rollup.sol

```diff
@@ -70,25 +75,17 @@ contract Rollup is IRollup {
 
     // Decode the cross-chain messages
     (bytes32 inHash,, bytes32[] memory l1ToL2Msgs, bytes32[] memory l2ToL1Msgs) =
-      MessagesDecoder.decode(_l2Block[HeaderDecoder.BLOCK_HEADER_SIZE:]);
-
-    bytes32 publicInputHash =
-      _computePublicInputHash(_l2Block[:HeaderDecoder.BLOCK_HEADER_SIZE], txsHash, inHash);
-
-    // @todo @LHerskind Proper genesis state. If the state is empty, we allow anything for now.
-    // TODO(#3936): Temporarily disabling this because L2Block encoding has not yet been updated.
-    // if (rollupStateHash != bytes32(0) && rollupStateHash != oldStateHash) {
-    //   revert Errors.Rollup__InvalidStateHash(rollupStateHash, oldStateHash);
-    // }
+      MessagesDecoder.decode(_body);
 
     bytes32[] memory publicInputs = new bytes32[](1);
-    publicInputs[0] = publicInputHash;
+    publicInputs[0] = _computePublicInputHash(_header, txsHash, inHash);
 
     if (!VERIFIER.verify(_proof, publicInputs)) {
       revert Errors.Rollup__InvalidProof();
     }
 
-    rollupStateHash = newStateHash;
+    // TODO: @benejsan Manually extracting the root here is ugly. TODO: Re-think how to represent archive snap.
+    archive = bytes32(_archive[:0x20]);
```

Comment:
Yes, we could pack it with `lastBlockTs` and it would be quite nice to have the ability to easily fetch the latest block number from contract.

>  for having an array of all the block archive etc, you can get away with not needing next available
@LHerskind how do you man this? I was thinking about just storing as a normal storage value packed with `lastBlockTs`.

Reply:
Recall when we were speaking about downloading a state snapshot such that you could sync really quickly.

So to make sure that the thing is valid, you have to ensure that it agrees with history, e.g., that the state actually matches the real state. The easiest way to do this (if you agree with the L1 contract for the canonical) you want to check that you agree with the contract. But if you are only having the last, you might have some troubles because it could have updated since the snapshot (this is very likely as we don't have state snapshots shared every block). 
This means that you would either try to search back in blocks on L1 to find the matching or you should figure it out as you are searching for blocks. 

Say that we are instead using something like below:
```solidity
uint256 blockNumber
mapping(uint256 blockNumber => bytes32 archive) archives;

function process(...) {
  archives[blockNumber++] = archive;
}
```
Then it is almost instant to check that the snapshot is good, you just look up at that block and you are done. 

In practice it is mainly a bit of performance diff as you can just lookup in state instead of having to look for the block though, so might not be worth it.

Reply:
We could modify the L2BlockProcessed event to emit not only block number but the archive as well. That would be cheaper and also very nice to use given that the blockNumber is indexed. WDYT?

(Will rework it to store the blockNumber in storage.)

Reply:
Realized that we have the block number in the header so we don't really need to store it in storage.

Reply:
Not sure you can fully get away with not having it there. So, we have it there such that you can ensure the next block is the old + 1. 

When doing the insertion into the archive, I guess you are matching the parent to be a position old, and then you ensure that old + 1 is empty and then insert. 

Assuming that archives is filled with zeros. Then if jumping ahead, you need to make a header that is hashing to 0 in order to use it. And if going in the past, you fail because the old + 1 would not be 0.

Think the yellow paper also need a `-1` actually in:
```python
# Check that the parent is a valid parent
    assert merkle_inclusion(
        parent.hash(),
        parent_sibling_path,
        left.public_inputs.constants.global_variables.block_number,
        left.public_inputs.constants.last_archive.root
    )
```

It would mean that it fails due to a proof error instead of bad block so error message is less clear though.